Python code refactoring.

Gadgetoid opened the pull request:
I got a little carried away, so this all ended up in one single commit. Raising as a draft until I can unpick it into some reasonable suggestions. Though feel free to cherry pick what matters to you.

It's also completely untested, but is thoroughly linted so I hopefully haven't done anything completely silly.

Some key changes:

* Rewrite terminal ID handling (picking up your todo) to use the stl urllib.parse, then split the path on "/" and grab the last element
* Move terminal ID creation into Terminal, and make it based on a UUID so it can't get potluck conflicting IDs (random has a tendency of not being random enough)
* Rewrite started, completed and alive to be a little more concise
* Try to use an explicit `encode("utf-8")` instead of a blunt convert to `bytes()`
* Try to make `get_terminal_env` a little cleaner
* Remove `Optional` from `cmdline` class variable, since `cmdline` is used regardless and could be undefined (raising an `AttributeError`)
* Rename from variables away from Python builtins (`min`, `id`) to avoid future mishaps

Alex4386 commented:
Will look into it! Thanks for the PR!

Alex4386 commented on defaults/py_modules/decky_terminal/__init__.py:
```diff
@@ -26,36 +31,31 @@ def __init__(self) -> None:
     def get_config_filename(self) -> str:
-        return DECKY_PLUGIN_SETTINGS_DIR+os.sep+"config.json"
-    
+        return os.path.join(DECKY_PLUGIN_SETTINGS_DIR, "config.json")
```
Nice change 👍🏼

Alex4386 commented on defaults/py_modules/decky_terminal/__init__.py:
```diff
@@ -81,107 +81,104 @@ async def get_default_shell(self) -> str:
 
-        print('cmdline!!!!', cmdline)
+        if terminal_id is None:
+            terminal_id = f"term-{uuid.uuid4()}"
```
The terminal ID can remove `term-` prefix since uuid itself is unique enough.  
The previous implementation had terminal_id of `term_####` format due to readability of uninitialized terminal having too long name on frontend. (the frontend fallbacks to terminal `id` when there is no `title` present)  

This can be mitigated via setting `title` of the terminal as `term_{last 4 digits of uuid}` or other `"user-identifiable measures"`

Alex4386 replied:
Will implement that on frontend side after the merge. Thanks

Gadgetoid replied:
Oh that's a good idea!

Alex4386 commented on defaults/py_modules/decky_terminal/common.py:
```diff
@@ -30,18 +31,18 @@ async def write_file(cls, filename: str, content: str) -> bool:
                 prev[i] = v
             else:
-                if type(prev.get(i)) is dict and type(v) is dict:
+                if isinstance(prev.get(i), dict) and isinstance(v, dict):
```
Nice cleanup 👍🏼

Alex4386 commented on defaults/py_modules/decky_terminal/terminal.py:
```diff
@@ -105,37 +104,39 @@ async def _process_subscriber(self, ws: WebSocketServerProtocol):
-        result["LINES"] = str(self.rows)
-        result["COLUMNS"] = str(self.cols)
-        result["XDG_RUNTIME_DIR"] = "/run/user/"+str(os.getuid())
+        result = dict(os.environ)  # dict(os.environ) == dict(**os.environ)
```
Nice cleanup 👍🏼

Alex4386 commented:
@Gadgetoid Looks OK! Since it is marked as WIP, Is it OK to merge into `main`?

Gadgetoid commented:
> @Gadgetoid Looks OK! Since it is marked as WIP, Is it OK to merge into `main`?

It remains *completely untested*, outside of some basic linting and code quality checks. 

If it works as you expect then it's probably okay to merge but you might want some more concise, individual commits (other than "Untested code savagery" in case any specific change breaks your code- since it'll be difficult to roll that one thing back amidst this mess. That's entirely up to you, though!

Might take me a while to come up with cleaner commits, though, since I'm buried in work and, wow, I really *did* get carried away here.

Gadgetoid committed:
main.py: drop unused imports.

Gadgetoid committed:
main.py: remove exception handling around decky_terminal import.

DeckyTerminal is used later in the code, so this basic exception
trap may print an ImportError but the code will later break with an
unhandled exception.

Gadgetoid committed:
main.py: reformat with black.

Gadgetoid committed:
main.py: delete dead code and comments.

Dropped commented-out code, unused test code and noisy comments
from the plugin template.

Gadgetoid committed:
decky_terminal/__init__.py: sort and prune imports.

Remove unused imports.

Sort imports with isort.

Gadgetoid committed:
decky_terminal/__init__.py: drop unused variable.

Variable "query_string" was asisgned but unused.

Gadgetoid committed:
decky_terminal/__init__.py: fix type comparison to isinstance.

Gadgetoid committed:
decky_terminal/common.py: prune and sort imports.

Remove unused imports.

Sort imports with isort.

Gadgetoid committed:
decky_terminal/common.py: fix type comparisons to isinstance.

Gadgetoid committed:
decky_terminal/terminal.py: sort and prune imports.

Remove unused import.

Sort imports with isort.

Gadgetoid committed:
decky_terminal/terminal.py: Fix type comparison and True comparison.

Gadgetoid committed:
decky_terminal/terminal.py: remove commented-out code.

Gadgetoid committed:
decky_terminal/terminal.py: reformat with black.

Gadgetoid committed:
decky_terminal/__init__.py: reformat with black.

Gadgetoid committed:
decky_terminal/common.py: reformat with black.

Gadgetoid committed:
decky_terminal/terminal.py: remove Optional qualifier on cmdline.

This variable is used in _start_process but Optional here does
not guarantee that it exists and is None.

Gadgetoid committed:
decky_terminal/terminal.py: do not shadow builtin min.

Gadgetoid committed:
decky_terminal/terminal.py: slightly more concise title empty check.

Gadgetoid committed:
decky_terminal/terminal.py: drop redundant spread operator.

dict(os.environ) == dict(**os.environ)

Gadgetoid committed:
decky_terminal/terminal.py: slightly less verbose env dict.

Gadgetoid committed:
decky_terminal/terminal.py: cleaner status checks.

Drop double-negatives and indirection to make this code easier to parse.

Gadgetoid committed:
all: refactor id to terminal_id.

Makes code easier to read, search and refactor in future.

Gadgetoid committed:
main.py: slightly less obtuse get_terminals.

Gadgetoid committed:
decky_terminal/__init__.py: use urlparse to parse out the terminal_id.

urllib is in the Python standard library, so we can rely on it.

Gadgetoid committed:
decky_terminal/__init__.py: clean up flow in get_shells.

Remove the redundant exception raise/catch serving as a goto.

Only return the default value in one place.

Gadgetoid committed:
decky_terminal/__init__.py: use os.path.join for get_config_filename.

Gadgetoid committed:
decky_terminal/__init__.py: simplify _is_unix_shell_path logic.

path.startswith("/") implies that len(path) > 0 and that path is not a space
or whitespace.

Gadgetoid committed:
generate a uuid instead of a random number for terminal_id.

Gadgetoid commented:
Just force pushed a, hopefully, verbose commit history. A diff of the changes versus my previous monolithic commit reveals the following differences.

TLDR: I've caught a couple of additional idiomatic python things, removed some logical redundancy and been a little more brutal removing dead code and let `black` have its way wrapping lines I don't necessarily agree with.

⚠️ Oh! And *still untested!*- I should probably rectify that.

```diff
diff --git a/defaults/py_modules/decky_terminal/__init__.py b/defaults/py_modules/decky_terminal/__init__.py
index 23b5916..f9199d8 100644
--- a/defaults/py_modules/decky_terminal/__init__.py
+++ b/defaults/py_modules/decky_terminal/__init__.py
@@ -44,13 +44,13 @@ class DeckyTerminal:
             if data is not None:
                 shells = data.splitlines()
                 shells = list(filter(self._is_unix_shell_path, shells))
-                if len(shells) > 0:
+                if shells:
                     return shells
 
             return ["/bin/sh"]
 
     def _is_unix_shell_path(self, path: str) -> bool:
-        return len(path) > 0 and path.startswith("/") and not path.isspace()
+        return path.startswith("/")
 
     # CONFIG ================================================
     def get_config_filename(self) -> str:
@@ -122,11 +122,11 @@ class DeckyTerminal:
         if cmdline is None:
             cmdline = await self.get_default_shell()
 
-        if terminal_id is None:
-            terminal_id = f"term-{uuid.uuid4()}"
-
         flags = await self._get_terminal_flags()
 
+        if terminal_id is None:
+            terminal_id = str(uuid.uuid4())
+
         if self._terminal_sessions.get(terminal_id) is None:
             self._terminal_sessions[terminal_id] = Terminal(cmdline, **flags)
 
diff --git a/defaults/py_modules/decky_terminal/terminal.py b/defaults/py_modules/decky_terminal/terminal.py
index 4f933cb..e60b4ea 100644
--- a/defaults/py_modules/decky_terminal/terminal.py
+++ b/defaults/py_modules/decky_terminal/terminal.py
@@ -35,7 +35,8 @@ class Terminal:
     _title_cache: bytes = b""
 
     def __init__(self, cmdline: str, is_shell: bool = True, **kwargs):
-        self.cmdline = cmdline  # cmdline can't meaningfully be None or undefined since it must be available for _start_process
+        self.cmdline = cmdline  # TODO: maybe raise ValueError? cmdline can't meaningfully be None or undefined since it must be available for _start_process
+
         self.is_shell = is_shell
         self.buffer = collections.deque([], maxlen=4096)
         self.subscribers = []
@@ -89,11 +90,15 @@ class Terminal:
 
         if self.master_fd is not None:
             new_size = struct.pack("HHHH", rows, cols, 0, 0)
-            await Common._run_async(fcntl.ioctl, self.master_fd, termios.TIOCSWINSZ, new_size)
+            await Common._run_async(
+                fcntl.ioctl, self.master_fd, termios.TIOCSWINSZ, new_size
+            )
 
         if self.slave_fd is not None:
             new_size = struct.pack("HHHH", rows, cols, 0, 0)
-            await Common._run_async(fcntl.ioctl, self.slave_fd, termios.TIOCSWINSZ, new_size)
+            await Common._run_async(
+                fcntl.ioctl, self.slave_fd, termios.TIOCSWINSZ, new_size
+            )
 
     # WORKERS ==============================================
     async def _process_subscriber(self, ws: WebSocketServerProtocol):
@@ -115,7 +120,7 @@ class Terminal:
 
     # PROCESS CONTROL =======================================
     def get_terminal_env(self):
-        result = dict(os.environ)  # dict(os.environ) == dict(**os.environ)
+        result = dict(os.environ)
 
         result.update({
             "TERM": "xterm-256color",
@@ -136,7 +141,6 @@ class Terminal:
 
     async def _start_process(self):
         self.master_fd, self.slave_fd = pty.openpty()
-        # self._set_pty_settings()
 
         await self._change_pty_size(self.rows, self.cols)
         self.process = await asyncio.create_subprocess_exec(
@@ -146,7 +150,6 @@ class Terminal:
             stderr=self.slave_fd,
             stdin=self.slave_fd,
             env=self.get_terminal_env(),
-            # creationflags=subprocess.CREATE_NO_WINDOW,
             cwd=os.getenv("HOME"),
         )
 
@@ -215,7 +218,9 @@ class Terminal:
         await Common._run_async(os.fsync, self.master_fd)
 
     async def _read_output(self) -> bytes:
-        output = await Common._run_async(os.read, self.master_fd, self._calculate_sync_size())
+        output = await Common._run_async(
+            os.read, self.master_fd, self._calculate_sync_size()
+        )
         if len(output) > 0:
             self._put_buffer(output)
             await self.broadcast_subscribers(output)
diff --git a/main.py b/main.py
index 6678be7..11942d1 100644
--- a/main.py
+++ b/main.py
@@ -1,12 +1,5 @@
 from typing import List, Optional
-
-import decky_plugin
-
-try:
-    from decky_terminal import DeckyTerminal
-except Exception as e:
-    decky_plugin.logger.info("[DeckyTerminal] Import Failed:", type(e), e)
-    raise e
+from decky_terminal import DeckyTerminal
 
 
 class Plugin:
@@ -24,24 +17,20 @@ class Plugin:
 
         for terminal_id, terminal in terminals.items():
             result = terminal.serialize()
-            output.append(
-                dict(
-                    id=terminal_id,
-                    **result,
-                )
-            )
+            result["id"] = terminal_id  # TODO: should a terminal own its ID?
+            output.append(result)
 
         return output
 
-    async def get_terminal(self, id: str) -> Optional[dict]:
-        terminal = Plugin.decky_terminal.get_terminal(id)
+    async def get_terminal(self, terminal_id: str) -> Optional[dict]:
+        terminal = Plugin.decky_terminal.get_terminal(terminal_id)
         if terminal is None:
             return None
 
         return terminal.serialize()
 
-    async def set_terminal_title(self, id: str, title: str) -> bool:
-        Plugin.decky_terminal.set_terminal_title(id, title)
+    async def set_terminal_title(self, terminal_id: str, title: str) -> bool:
+        Plugin.decky_terminal.set_terminal_title(terminal_id, title)
 
     async def create_terminal(self, terminal_id=None) -> bool:
         await Plugin.decky_terminal.create_terminal(terminal_id)
@@ -54,18 +43,15 @@ class Plugin:
         except:
             return False
 
-    async def change_terminal_window_size(
-        self, terminal_id, rows: int, cols: int
-    ) -> bool:
+    async def change_terminal_window_size(self, terminal_id, rows: int, cols: int) -> bool:
         try:
             terminal = Plugin.decky_terminal.get_terminal(terminal_id)
             if terminal is not None:
                 await terminal.change_window_size(rows, cols)
                 return True
+            return False
         except:
-            pass
-
-        return False
+            return False
 
     async def get_config(self) -> str:
         return await Plugin.decky_terminal.get_config()
@@ -90,10 +76,3 @@ class Plugin:
 
     async def _migration(self):
         pass
-
-    async def _internal_test(self):
-        await self._main()
-        decky_plugin.logger.info(Plugin.decky_terminal.get_server_port())
-
-        await Plugin.decky_terminal.create_terminal("testterm")
-        decky_plugin.logger.info("testterm created.")

```

Gadgetoid commented:
Okay *now* I have done some cursory testing and quickly rebased out a bug in my urlparse code (`[:-1]` where I should have used `[-1]` ... that one always gets me).

Gadgetoid committed:
.vscode/setup.sh: add platform fixes from template.

Gadgetoid committed:
convert uuid to phonetic title.

Use a crude nato alphabet and number lookup to convert the first four
digits of the uuid into a four word title.

Terminals will be given more memorable names, such as NinerThreeSixEcho
or OneFiveBravoFive etc.

Gadgetoid commented:
Inspired by your idea for terminal naming, I wondered what it would look like if I converted the uuid to a phonetic alphabet. I chose nato arbitrarily, but you could get fun with it-

![decky-terminal-screenshot](https://github.com/Alex4386/decky-terminal/assets/1746967/48e12528-f236-4984-8b88-de46be0df2d1)

Alex4386 commented:
> Inspired by your idea for terminal naming, I wondered what it would look like if I converted the uuid to a phonetic alphabet. I chose nato arbitrarily, but you could get fun with it-
> 
> ![decky-terminal-screenshot](https://private-user-images.githubusercontent.com/1746967/308748268-48e12528-f236-4984-8b88-de46be0df2d1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkxOTM2MjcsIm5iZiI6MTcwOTE5MzMyNywicGF0aCI6Ii8xNzQ2OTY3LzMwODc0ODI2OC00OGUxMjUyOC1mMjM2LTQ5ODQtOGI4OC1kZTQ2YmUwZGYyZDEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIyOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMjlUMDc1NTI3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9M2EyZDBjZjkyZGYwOTExMzY5ZWVhY2QwOGFmYzFkNzU5ZjgzYTRhNzE5OGY5NzUyMjg4NTE4NWRlNWY5NDk4MSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.RXCBRRUh0g5EzIHb9glGRm9Nt4oywVf48iVsiFXm3tQ)
  
ICAO Phonetic Alphabet seems like a fun way of representing UUIDs! And also, thanks for testing. I was unable to test your branch due to other projects.  
If you think it's ready to merge, feel free to ping me!

Alex4386 commented:
> TLDR: I've caught a couple of additional idiomatic python things, removed some logical redundancy and been a little more brutal removing dead code and let `black` have its way wrapping lines I don't necessarily agree with.
> 

You've might have seen my code, or from my other projects, I'm not really a Python person.  
That was the reason the code was not pythonic. Thanks for rectifying that. 😆  
  
Maybe I should setup Blake or other linting action later on 🤔

Gadgetoid commented:
I think the current in-vogue Python linting tool is “ruff”, I’ve been contemplating an upstream patch to the template to add that into CI since it can catch real actual bugs and not just syntax quirks.

“black” is a very opinionated reformatting tool that- depending on your preference- can either be useful, annoying or some mix of the two.

I’ll try and do some more in depth testing, currently the phonetic terminal titles don’t show up at the top of the individual terminals and I’m not a React person so I haven’t looked into that yet 🤣

I also had some ideas about making the Terminal class implicitly track open terminals/manage sessions, but that’s a little bit more ambitious and breaky!

Alex4386 commented:
OK, If you are ready. I'll merge it to the `main` branch.

Alex4386 commented:
Since this is no longer marked as WIP, I'll consider answer as Yes and merge.

Gadgetoid commented:
Sorry got distracted. It was ready enough! I'm a prod away if testers catch anything that needs fixing, just punt it over.

Alex4386 commented:
Getting ready with v0.4.0 release: https://github.com/SteamDeckHomebrew/decky-plugin-database/pull/580

Alex4386 commented:
Merged and Now available on stores!
The Acknowledgement page has been updated to mark your work! See [Line 23 on AcknowledgementPage.tsx](https://github.com/Alex4386/decky-terminal/blob/main/src/pages/settings/AcknowledgementPage.tsx#L23).

Thank you for your work! 😊 @Gadgetoid
